### PR TITLE
Add missing named optimizers

### PR DESCRIPTION
**Pull request by atanasster**
#### Description
added Adadelta (adadelta), Adamax (adamax), Momentum(momentum) as named optimizers to Layers API.
defaults taken from https://keras.io/optimizers/
---
##### For repository owners only:
FEATURE

PR for issue #401

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/229)

### CODE REVIEW

**Comment by atanasster** [comment]
I signed it!

**Comment by bileschi** [comment]
This is great!  Just a few small nits.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[src/optimizers.ts, line 28 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-LES0xgQ0uj2gB9rssnQ:-LES0xgQ0uj2gB9rssnR:bl8r3ma) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/df0b7f5377e4975b5e51c3d61bb03380c782e1cd/src/optimizers.ts#L28)):*
> ```XML
> // change based on (de)serialization requirements.
> export function getOptimizer(identifier: string): Optimizer {
>   const optimizerMap: { [optimizerName: string]: () => Optimizer } = {
> ```

Please keep this list alphabetized, it helps reader when they're looking for a value.  Thanks

---

*[src/optimizers_test.ts, line 16 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-LES0qrq4CWF5TgcMdTe:-LES0qrq4CWF5TgcMdTf:b-vi7ryj) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/df0b7f5377e4975b5e51c3d61bb03380c782e1cd/src/optimizers_test.ts#L16)):*
> ```XML
> 
> // tslint:disable:max-line-length
> import { AdagradOptimizer, AdamOptimizer, RMSPropOptimizer, SGDOptimizer, AdadeltaOptimizer, AdamaxOptimizer, MomentumOptimizer } from '@tensorflow/tfjs-core';
> ```

These should be alphabetized

---

*[src/optimizers_test.ts, line 18 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-LES1AvM4fTuYVklWYbB:-LES1AvM4fTuYVklWYbC:b-z8ol2u) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/df0b7f5377e4975b5e51c3d61bb03380c782e1cd/src/optimizers_test.ts#L18)):*
> ```XML
> import { AdagradOptimizer, AdamOptimizer, RMSPropOptimizer, SGDOptimizer, AdadeltaOptimizer, AdamaxOptimizer, MomentumOptimizer } from '@tensorflow/tfjs-core';
> 
> import { getOptimizer } from './optimizers';
> ```

Based on this added spacing, I think the style preferences in your IDE are conflicting with the style preferences for this project.  Can you check?  Thanks.

---

*[src/optimizers_test.ts, line 63 at r1](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-LES0dyD7Jgfpi-Wtyg9:-LES0dyD7Jgfpi-WtygA:b-jvchbp) ([raw file](https://github.com/tensorflow/tfjs-layers/blob/df0b7f5377e4975b5e51c3d61bb03380c782e1cd/src/optimizers_test.ts#L63)):*
<details><summary><i alt="summary">Quoted 26 lines of code…</i></summary>

> ```
>  
>   it(`can instantiate Adadelta`, () => {
>     const optimizer = getOptimizer('Adadelta');
>     expect(optimizer instanceof AdadeltaOptimizer).toBe(true);
>   });
>   it(`can instantiate adadelta`, () => {
>     const optimizer = getOptimizer('adadelta');
>     expect(optimizer instanceof AdadeltaOptimizer).toBe(true);
>   });
>  
>   it(`can instantiate Adamax`, () => {
>     const optimizer = getOptimizer('Adamax');
>     expect(optimizer instanceof AdamaxOptimizer).toBe(true);
>   });
>   it(`can instantiate adamax`, () => {
>     const optimizer = getOptimizer('adamax');
>     expect(optimizer instanceof AdamaxOptimizer).toBe(true);
>   });
>   it(`can instantiate Momentum`, () => {
>     const optimizer = getOptimizer('Momentum');
>     expect(optimizer instanceof MomentumOptimizer).toBe(true);
>   });
>   it(`can instantiate momentum`, () => {
>     const optimizer = getOptimizer('momentum');
>     expect(optimizer instanceof MomentumOptimizer).toBe(true);
>   });
> ```
</details>

Please move these above the test for "non-existent optimizer", so that the non-existent one is last.

---


*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-:-LES12JN3sn8eaOA1_yv:bmy15zh)*

**Comment by atanasster** [comment]
Thanks for the notes, i am still learning your stack :) will update the PR

**Comment by atanasster** [comment]
@bileschi - entered a new issue for the formatting, probably better to have the setting in GitHub so new contributors can start editing right away without misformatting for code

File: src/optimizers.ts [hunk 1]
```diff
@@ -27,12 +27,18 @@ import {ValueError} from './errors';
 export function getOptimizer(identifier: string): Optimizer {
   const optimizerMap: {[optimizerName: string]: () => Optimizer} = {
     'Adagrad': () => train.adagrad(.01),
+    'Adadelta': () => train.adadelta(1.0, 0.95, K.epsilon()),
     'Adam': () => train.adam(.001, .9, .999, K.epsilon()),
+    'Adamax': () => train.adamax(0.002, .9, .999, K.epsilon(), 0.0),
+    'Momentum': () => train.momentum(0.01, 0.0, false),
```
Comment (ericdnielsen) [inline src/optimizers.ts]: I think we want to remove the Momentum one.  All the other ones are cross-compatible with python and JS using the JSON serialization format with agreed upon symbol names.  Momentum is not in Keras at present.  But is in tensorflow & tensorflow.js core.  However the default arguments don't appear to make sense (why create a momentum optimizer with momentum set to 0?)

Comment (atanasster) [reply src/optimizers.ts]: I took the defaults for Momentum from https://keras.io/optimizers/ SGD Optimizer, where momentum is 0.0. 

Should I remove Momentum completely?

Comment (bileschi) [reply src/optimizers.ts]: Yes, please remove momentum.  This way users won't be confused if they serialized a model built in tfjs, and it then crashes in python keras.
thanks!

Comment (atanasster) [reply src/optimizers.ts]: thanks ! removed now Momentum

**Comment by caisq** [comment]
:lgtm_strong:

Thanks, @atanasster !

---

Review status: :shipit: complete! 1 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-:-LEkOGAkBvlPVlJYTSQ5:b8tqlvb)*

**Comment by bileschi** [comment]
:lgtm_strong:

---

Review status: :shipit: complete! 2 of 1 LGTMs obtained

---



*Comments from [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs-layers/229#-:-LEkPBTiCxPJ3tI63whP:b-jbsf8z)*

**Comment by caisq** [comment]
@atanasster I've pulled master into your PR. As soon as the tests pass, we can merge this PR.

**Pull request merged**